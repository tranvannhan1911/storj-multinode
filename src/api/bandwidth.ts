// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

import { APIClient } from '@/api/index';
import { BandwidthRollup, BandwidthTraffic, Egress, Ingress } from '@/bandwidth';

/**
 * Client for bandwidth controller of MND api.
 */
export class BandwidthClient extends APIClient {
    private readonly ROOT_PATH: string = '/api/v0/bandwidth';

    /**
     * Returns bandwidth information for selected node and satellite if any.
     *
     * @throws {@link BadRequestError}
     * This exception is thrown if the input is not a valid.
     *
     * @throws {@link UnauthorizedError}
     * Thrown if the auth cookie is missing or invalid.
     *
     * @throws {@link InternalError}
     * Thrown if something goes wrong on server side.
     */
    public async fetch(satelliteId: string | null, nodeId: string | null): Promise<BandwidthTraffic> {
        let path = `${this.ROOT_PATH}`;

        if (!satelliteId && !nodeId) {
            path += '/';
        }

        if (satelliteId) {
            path += `/satellites/${satelliteId}`;
        }

        if (nodeId) {
            path += `/${nodeId}`;
        }

        const response = await this.http.get(path);

        if (!response.ok) {
            await this.handleError(response);
        }

        const traffic = await response.json();
        const daily = traffic.bandwidthDaily || [];

        return new BandwidthTraffic(
            daily.map(daily => new BandwidthRollup(
                new Egress(daily.egress.repair, daily.egress.audit, daily.egress.usage),
                new Ingress(daily.ingress.repair, daily.ingress.usage),
                daily.delete,
                new Date(daily.intervalStart),
            )),
            traffic.bandwidthSummary,
            traffic.egressSummary,
            traffic.ingressSummary,
        );
    }
}
