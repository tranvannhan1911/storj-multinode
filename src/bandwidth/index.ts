// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

/**
 * Represents by day and total bandwidth.
 */
export class BandwidthTraffic {
    public constructor(
        public bandwidthDaily: BandwidthRollup[] = [],
        public bandwidthSummary: number = 0,
        public egressSummary: number = 0,
        public ingressSummary: number = 0,
    ) {}
}

/**
 * Represents by day bandwidth.
 */
export class BandwidthRollup {
    public constructor(
        public egress: Egress = new Egress(),
        public ingress: Ingress = new Ingress(),
        public deletes: number = 0,
        public intervalStart: Date = new Date(),
    ) {}

    /**
     * Creates new empty instance of used bandwidth with defined date.
     * @param date - holds specific date of the month
     * @returns BandwidthUsed - new empty instance of used bandwidth with defined date
     */
    public static emptyWithDate(date: number): BandwidthRollup {
        const now = new Date();

        now.setUTCDate(date);
        now.setUTCHours(0, 0, 0, 0);

        return new BandwidthRollup(new Egress(0, 0, 0), new Ingress(0, 0), 0, now);
    }
}

/**
 * Stores info about storage node egress usage.
 */
export class Egress {
    public constructor(
        public repair: number = 0,
        public audit: number = 0,
        public usage: number = 0,
    ) {}
}

/**
 * Stores info about storage node ingress usage.
 */
export class Ingress {
    public constructor(
        public repair: number = 0,
        public usage: number = 0,
    ) {}
}
