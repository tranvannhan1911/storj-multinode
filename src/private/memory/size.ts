// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

/**
 * Base10 sizes.
 */
export enum SizeBreakpoints {
    KB = 1e3,
    MB = 1e6,
    GB = 1e9,
    TB = 1e12,
    PB = 1e15,
    EB = 1e18,
}

/**
 * Size class contains size related functionality such as convertation.
 */
export class Size {
    /**
     * Base10String converts size to a string using base-10 prefixes.
     * @param size in bytes
     */
    public static toBase10String(size: number): string {
        const decimals = 2;

        const _size = Math.abs(size);

        switch (true) {
        case _size >= SizeBreakpoints.EB * 2 / 3:
            return `${parseFloat((size / SizeBreakpoints.EB).toFixed(decimals))}EB`;
        case _size >= SizeBreakpoints.PB * 2 / 3:
            return `${parseFloat((size / SizeBreakpoints.PB).toFixed(decimals))}PB`;
        case _size >= SizeBreakpoints.TB * 2 / 3:
            return `${parseFloat((size / SizeBreakpoints.TB).toFixed(decimals))}TB`;
        case _size >= SizeBreakpoints.GB * 2 / 3:
            return `${parseFloat((size / SizeBreakpoints.GB).toFixed(decimals))}GB`;
        case _size >= SizeBreakpoints.MB * 2 / 3:
            return `${parseFloat((size / SizeBreakpoints.MB).toFixed(decimals))}MB`;
        case _size >= SizeBreakpoints.KB * 2 / 3:
            return `${parseFloat((size / SizeBreakpoints.KB).toFixed(decimals))}KB`;
        default:
            return `${size ? size.toFixed(decimals) : size}B`;
        }
    }
}
