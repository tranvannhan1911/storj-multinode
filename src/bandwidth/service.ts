// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

import { BandwidthClient } from '@/api/bandwidth';
import { BandwidthTraffic } from '@/bandwidth/index';

/**
 * exposes all bandwidth related logic
 */
export class Bandwidth {
    private readonly bandwidth: BandwidthClient;

    public constructor(bandwidth: BandwidthClient) {
        this.bandwidth = bandwidth;
    }

    /**
     * returns bandwidth for selected satellite and node if any.
     *
     * @throws {@link BadRequestError}
     * This exception is thrown if the input is not a valid.
     *
     * @throws {@link UnauthorizedError}
     * Thrown if the auth cookie is missing or invalid.
     *
     * @throws {@link InternalError}
     * Thrown if something goes wrong on server side.
     */
    public async fetch(satelliteId: string | null, nodeId: string | null): Promise<BandwidthTraffic> {
        return await this.bandwidth.fetch(satelliteId, nodeId);
    }
}
