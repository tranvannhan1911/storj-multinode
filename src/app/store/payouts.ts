// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

import { ActionContext, ActionTree, GetterTree, Module, MutationTree } from 'vuex';

import { RootState } from '@/app/store/index';
import { monthNames } from '@/app/types/date';
import { Expectation, HeldAmountSummary, NodePayouts, PayoutsSummary, Paystub } from '@/payouts';
import { Payouts } from '@/payouts/service';

/**
 * PayoutsState is a representation of payouts module state.
 */
export class PayoutsState {
    public summary: PayoutsSummary = new PayoutsSummary();
    public selectedPayoutPeriod: string | null = null;
    public selectedNodePayouts: NodePayouts = new NodePayouts();
    public selectedNodeExpectations: Expectation = new Expectation();
    public totalExpectations: Expectation = new Expectation();
}

/**
 * PayoutsModule is a part of a global store that encapsulates all payouts related logic.
 */
export class PayoutsModule implements Module<PayoutsState, RootState> {
    public readonly namespaced: boolean;
    public readonly state: PayoutsState;
    public readonly getters?: GetterTree<PayoutsState, RootState>;
    public readonly actions: ActionTree<PayoutsState, RootState>;
    public readonly mutations: MutationTree<PayoutsState>;

    private readonly payouts: Payouts;

    public constructor(payouts: Payouts) {
        this.payouts = payouts;

        this.namespaced = true;
        this.state = new PayoutsState();

        this.mutations = {
            setSummary: this.setSummary,
            setPayoutPeriod: this.setPayoutPeriod,
            setTotalExpectation: this.setTotalExpectation,
            setNodeTotals: this.setNodeTotals,
            setNodePaystub: this.setNodePaystub,
            setNodeHeldHistory: this.setNodeHeldHistory,
            setCurrentNodeExpectations: this.setCurrentNodeExpectations,
        };

        this.actions = {
            summary: this.summary.bind(this),
            expectations: this.expectations.bind(this),
            paystub: this.paystub.bind(this),
            nodeTotals: this.nodeTotals.bind(this),
            heldHistory: this.heldHistory.bind(this),
        };

        this.getters = {
            periodString: this.periodString,
        };
    }

    // Mutations
    /**
     * setSummary mutation will set payouts summary state.
     * @param state - state of the module.
     * @param summary - payouts summary information depends on selected time and satellite.
     */
    public setSummary(state: PayoutsState, summary: PayoutsSummary): void {
        state.summary = summary;
    }

    /**
     * setPayoutPeriod mutation will save selected period to store.
     * @param state
     * @param period representation of month and year
     */
    public setPayoutPeriod(state: PayoutsState, period: string | null): void {
        state.selectedPayoutPeriod = period;
    }

    /**
     * setTotalExpectation mutation will set total payouts expectation for all nodes.
     * @param state - state of the module.
     * @param expectations - payouts summary information depends on selected time and satellite.
     */
    public setTotalExpectation(state: PayoutsState, expectations: Expectation): void {
        state.totalExpectations = expectations;
    }

    /**
     * setNodeTotals mutation will total payout information for selected node to store.
     * @param state
     * @param paystub for all time
     */
    public setNodeTotals(state: PayoutsState, paystub: Paystub): void {
        state.selectedNodePayouts = { ...state.selectedNodePayouts, totalPaid: paystub.distributed, totalEarned: paystub.paid, totalHeld: paystub.held };
    }

    /**
     * setNodePaystub mutation will save paystub for selected node, satellite and period to store.
     * @param state
     * @param paystub
     */
    public setNodePaystub(state: PayoutsState, paystub: Paystub): void {
        state.selectedNodePayouts = { ...state.selectedNodePayouts, paystubForPeriod: paystub };
    }

    /**
     * setNodeHeldHistory mutation will save held history for selected node to store.
     * @param state
     * @param heldHistory
     */
    public setNodeHeldHistory(state: PayoutsState, heldHistory: HeldAmountSummary[]): void {
        state.selectedNodePayouts = { ...state.selectedNodePayouts, heldHistory };
    }

    /**
     * setCurrentNodeExpectations mutation will save paystub for selected node, satellite and period to store.
     * @param state
     * @param expectations
     */
    public setCurrentNodeExpectations(state: PayoutsState, expectations: Expectation): void {
        state.selectedNodePayouts = { ...state.selectedNodePayouts, expectations };
    }

    // Actions
    /**
     * summary action loads payouts summary information.
     * @param ctx - context of the Vuex action.
     */
    public async summary(ctx: ActionContext<PayoutsState, RootState>): Promise<void> {
        const selectedSatelliteId = ctx.rootState.nodes.selectedSatellite ? ctx.rootState.nodes.selectedSatellite.id : null;
        const summary = await this.payouts.summary(selectedSatelliteId, ctx.state.selectedPayoutPeriod);

        ctx.commit('setSummary', summary);
    }

    /**
     * expectations action loads payouts total or by node payout expectation information.
     * @param ctx - context of the Vuex action.
     * @param nodeId - node id.
     */
    public async expectations(ctx: ActionContext<PayoutsState, RootState>, nodeId: string | null): Promise<void> {
        const expectations = await this.payouts.expectations(nodeId);

        ctx.commit(`${nodeId ? 'setCurrentNodeExpectations' : 'setTotalExpectation'}`, expectations);
    }

    /**
     * paystub action loads payouts information for table for selected node.
     * @param ctx - context of the Vuex action.
     * @param nodeId
     */
    public async paystub(ctx: ActionContext<PayoutsState, RootState>, nodeId: string): Promise<void> {
        const selectedSatelliteId = ctx.rootState.nodes.selectedSatellite ? ctx.rootState.nodes.selectedSatellite.id : null;
        const paystub = await this.payouts.paystub(selectedSatelliteId, ctx.state.selectedPayoutPeriod, nodeId);

        ctx.commit('setNodePaystub', paystub);
    }

    /**
     * nodeTotals action loads total payouts information for selected node.
     * @param ctx - context of the Vuex action.
     * @param nodeId
     */
    public async nodeTotals(ctx: ActionContext<PayoutsState, RootState>, nodeId: string): Promise<void> {
        const paystub = await this.payouts.paystub(null, null, nodeId);

        ctx.commit('setNodeTotals', paystub);
    }

    /**
     * heldHistory action loads held history for selected node.
     * @param ctx - context of the Vuex action.
     * @param nodeId
     */
    public async heldHistory(ctx: ActionContext<PayoutsState, RootState>, nodeId: string): Promise<void> {
        const heldHistory = await this.payouts.heldHistory(nodeId);

        ctx.commit('setNodeHeldHistory', heldHistory);
    }

    // Getters
    /**
     * periodString is full name month and year representation of selected payout period.
     */
    public periodString(state: PayoutsState): string {
        if (!state.selectedPayoutPeriod) { return 'All time'; }

        const splittedPeriod = state.selectedPayoutPeriod.split('-');
        const monthIndex = parseInt(splittedPeriod[1]) - 1;
        const year = splittedPeriod[0];

        return `${monthNames[monthIndex]}, ${year}`;
    }
}
