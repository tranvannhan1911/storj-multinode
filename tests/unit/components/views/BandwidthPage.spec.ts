// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

import Vuex from 'vuex';
import { createLocalVue, shallowMount } from '@vue/test-utils';

import store from '../../mock/store';

import BandwidthPage from '@/app/views/bandwidth/BandwidthPage.vue';
import { BandwidthTraffic } from '@/bandwidth';
import { Size } from '@/private/memory/size';

const localVue = createLocalVue();

localVue.use(Vuex);

localVue.filter('bytesToBase10String', (amountInBytes: number): string => Size.toBase10String(amountInBytes));

const traffic = new BandwidthTraffic();

traffic.bandwidthSummary = 700000000;
traffic.egressSummary = 577700000000;
traffic.ingressSummary = 5000000;

describe('BandwidthPage', (): void => {
    it('renders correctly', (): void => {
        store.commit('bandwidth/populate', traffic);

        const wrapper = shallowMount(BandwidthPage, {
            store,
            localVue,
        });

        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly with egress chart', async(): Promise<void> => {
        const wrapper = shallowMount(BandwidthPage, {
            store,
            localVue,
        });

        await wrapper.findAll('.chart-container__title-area__chart-choice-item').at(1).trigger('click');

        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly with ingress chart', async(): Promise<void> => {
        const wrapper = shallowMount(BandwidthPage, {
            store,
            localVue,
        });

        await wrapper.findAll('.chart-container__title-area__chart-choice-item').at(2).trigger('click');

        expect(wrapper).toMatchSnapshot();
    });
});
