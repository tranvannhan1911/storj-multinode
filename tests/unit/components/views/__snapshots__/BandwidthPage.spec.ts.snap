// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BandwidthPage renders correctly 1`] = `
<div class="bandwidth">
  <h1 class="bandwidth__title">Bandwidth &amp; Disk</h1>
  <div class="bandwidth__dropdowns">
    <node-selection-dropdown-stub></node-selection-dropdown-stub>
    <satellite-selection-dropdown-stub></satellite-selection-dropdown-stub>
  </div>
  <div class="chart-container bandwidth-chart">
    <div class="chart-container__title-area">
      <p class="chart-container__title-area__title">Bandwidth Used This Month</p>
      <div class="chart-container__title-area__buttons-area"><button name="Show Bandwidth Chart" type="button" class="chart-container__title-area__chart-choice-item active">
          Bandwidth
        </button> <button name="Show Egress Chart" type="button" class="chart-container__title-area__chart-choice-item">
          Egress
        </button> <button name="Show Ingress Chart" type="button" class="chart-container__title-area__chart-choice-item">
          Ingress
        </button></div>
    </div>
    <p class="chart-container__amount"><b>0.7GB</b></p>
    <div onresize="recalculateChartDimensions()" class="chart-container__chart">
      <bandwidth-chart-stub width="0" height="0"></bandwidth-chart-stub>
    </div>
  </div>
  <section class="bandwidth__chart-area">
    <section class="chart-container">
      <div class="chart-container__title-area disk-space-title">
        <p class="chart-container__title-area__title">Average Disk Space Used This Month</p>
      </div>
      <p class="chart-container__amount disk-space-amount"><b>0B</b></p>
      <div onresize="recalculateChartDimensions()" class="chart-container__chart">
        <disk-space-chart-stub width="0" height="0"></disk-space-chart-stub>
      </div>
    </section>
    <section class="disk-stat-chart">
      <disk-stat-chart-stub></disk-stat-chart-stub>
    </section>
  </section>
</div>
`;

exports[`BandwidthPage renders correctly with egress chart 1`] = `
<div class="bandwidth">
  <h1 class="bandwidth__title">Bandwidth &amp; Disk</h1>
  <div class="bandwidth__dropdowns">
    <node-selection-dropdown-stub></node-selection-dropdown-stub>
    <satellite-selection-dropdown-stub></satellite-selection-dropdown-stub>
  </div>
  <div class="chart-container bandwidth-chart">
    <div class="chart-container__title-area">
      <p class="chart-container__title-area__title">Bandwidth Used This Month</p>
      <div class="chart-container__title-area__buttons-area"><button name="Show Bandwidth Chart" type="button" class="chart-container__title-area__chart-choice-item">
          Bandwidth
        </button> <button name="Show Egress Chart" type="button" class="chart-container__title-area__chart-choice-item active">
          Egress
        </button> <button name="Show Ingress Chart" type="button" class="chart-container__title-area__chart-choice-item">
          Ingress
        </button></div>
    </div>
    <p class="chart-container__amount"><b>577.7GB</b></p>
    <div onresize="recalculateChartDimensions()" class="chart-container__chart">
      <egress-chart-stub width="0" height="0"></egress-chart-stub>
    </div>
  </div>
  <section class="bandwidth__chart-area">
    <section class="chart-container">
      <div class="chart-container__title-area disk-space-title">
        <p class="chart-container__title-area__title">Average Disk Space Used This Month</p>
      </div>
      <p class="chart-container__amount disk-space-amount"><b>0B</b></p>
      <div onresize="recalculateChartDimensions()" class="chart-container__chart">
        <disk-space-chart-stub width="0" height="0"></disk-space-chart-stub>
      </div>
    </section>
    <section class="disk-stat-chart">
      <disk-stat-chart-stub></disk-stat-chart-stub>
    </section>
  </section>
</div>
`;

exports[`BandwidthPage renders correctly with ingress chart 1`] = `
<div class="bandwidth">
  <h1 class="bandwidth__title">Bandwidth &amp; Disk</h1>
  <div class="bandwidth__dropdowns">
    <node-selection-dropdown-stub></node-selection-dropdown-stub>
    <satellite-selection-dropdown-stub></satellite-selection-dropdown-stub>
  </div>
  <div class="chart-container bandwidth-chart">
    <div class="chart-container__title-area">
      <p class="chart-container__title-area__title">Bandwidth Used This Month</p>
      <div class="chart-container__title-area__buttons-area"><button name="Show Bandwidth Chart" type="button" class="chart-container__title-area__chart-choice-item">
          Bandwidth
        </button> <button name="Show Egress Chart" type="button" class="chart-container__title-area__chart-choice-item">
          Egress
        </button> <button name="Show Ingress Chart" type="button" class="chart-container__title-area__chart-choice-item active">
          Ingress
        </button></div>
    </div>
    <p class="chart-container__amount"><b>5MB</b></p>
    <div onresize="recalculateChartDimensions()" class="chart-container__chart">
      <ingress-chart-stub width="0" height="0"></ingress-chart-stub>
    </div>
  </div>
  <section class="bandwidth__chart-area">
    <section class="chart-container">
      <div class="chart-container__title-area disk-space-title">
        <p class="chart-container__title-area__title">Average Disk Space Used This Month</p>
      </div>
      <p class="chart-container__amount disk-space-amount"><b>0B</b></p>
      <div onresize="recalculateChartDimensions()" class="chart-container__chart">
        <disk-space-chart-stub width="0" height="0"></disk-space-chart-stub>
      </div>
    </section>
    <section class="disk-stat-chart">
      <disk-stat-chart-stub></disk-stat-chart-stub>
    </section>
  </section>
</div>
`;
